refactor/feat(codebase): Refactor API-compatible. Added many features.

xxxtonixxx opened the pull request:
In this refactor we have added new features

- [x] Created a better organization of files and types.
- [x] Created a way of communicate with our children components
- [x] An API more robust and open
- [x] Created external events to get changes from inputs (onChange, onBlur, onChange, onEnter, onEscape, onFocus, etc.). These events are shotcuts to angular-native events (ex. keypress.enter, keypress.escape), so you can bind those. However, this shorcuts provide you of state of input when this one changed.
- [x] New options to configure when an event is triggered (ex. hidebuttons, cancelOnEscape, saveOnEnter, saveOnBlur, etc.)
- [x] An internal service shared between root and child component. (this feature will be more use in the future, so it is WIP)
- [x] Interfaces to provide a specifig config to inputs (specific intellisense to input text, input select, etc.)
    
How works the new API to communicate the root component with their children?
When a child component changes its state, this one triggers an event and the root component get it.
Using this system of internal events we can provide external events (to the client) and trigger events such as onChange, onFocus, on Blur, etc.
A change of state can be: change of value, edit mode or disabled mode.

### **To do done before it can be merged:**
- [x] Better names (it's extremely important to choose this, once merged, change this will produce a breaking change), ex:
- into `types` folder the name of files is inconsistente with the name of types.
- the name of the event types (`InteralEvent`/`ExternalEvent`, `ExternalEvent` is exported like `InlineEvent`, is it a good name?).
- `InputTextComponent`, `InputSelectComponent`, etc. could be `TextComponent` only, and the config could be `TextConfig`
- [x] An option to configure if we want only the value changed or the state object (maybe `onlyValue` or `withState`). **Configured by default like the first one.**
- [x] Choose good paths from where import the API (ex, to import `InputTextConfig` the path could be `ng2-inline-editor/inputs` or import all from `ng2-inline-editor`)

Closes #38 #47 #30 #21 #24

Caballerog commented:
Should be interesting that you would retrieval the data in "value" property in a default configuration which it is relevant value in the most of use case. Furthermore, the API is not break change.

xxxtonixxx commented:
I have updated the PR's description to add a todo list and I have added your feedback 👍 .

It's important to close those points before merge, so I would appreciate to receive more feedback 😄

Caballerog commented:
Hi!

Tomorrow, I going to write some interesting features which would be in the
1.0.

Bye!
El El vie, 19 may 2017 a las 13:17, Toni <notifications@github.com>
escribió:

> I have updated the PR's description to add a todo list and I have added
> your feedback 👍 .
>
> It's important to close those points before merge, so I would appreciate
> to receive more feedback 😄
>
> —
> You are receiving this because you commented.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/Caballerog/ng2-inline-editor/pull/57#issuecomment-302676780>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AIZj3y6lm-m0vf8jiphwxo0gN4XtBfJIks5r7XpUgaJpZM4Nbukv>
> .
>
-- 
Enviado desde el teléfono móvil, disculpe las faltas.

xxxtonixxx commented:
The `onlyValue` option is added with https://github.com/Caballerog/ng2-inline-editor/pull/57/commits/dbf931ff64f13a43f17333965b35d89eca6e930d and it's `true` by default 👍

xxxtonixxx commented:
In the last commit I have done changes:
- `InlineEvent` -> `InlineEditorEvent`
- `InlineError` -> `InlineEditorError`
- `ExternalEvent` is defined but never used, it's always used its alias (`InlineEditorEvent`)
- File names have been renamed (`src/types/inline-*.ts` -> `src/types/inline-editor-*.ts`)
- Some changes more, but I don't remember them 😅 

I think it can be merged 👍

Caballerog commented:
This commit has several break changes in the API!

The error handler is different, now you should use `(onError)="handleError()" instead of [fnError]="handleError"`.

Now you could configure your component using the property: `config`.
